### PR TITLE
Trivial refactoring in scala / library

### PR DESCRIPTION
**Pull request by jxcoder**
Refactored:
- [scala.collection.convert.Wrappers](//github.com/jxcoder/scala/blob/2.11.x/src/library/scala/collection/convert/Wrappers.scala#L422)
- [scala.collection.concurrent.MainNode](//github.com/jxcoder/scala/blob/2.11.x/src/library/scala/beans/ScalaBeanInfo.scala)
- [scala.collection.concurrent.INodeBase](//github.com/jxcoder/scala/blob/2.11.x/src/library/scala/collection/concurrent/INodeBase.java)
- [scala.collection.concurrent.CNodeBase](//github.com/jxcoder/scala/blob/2.11.x/src/library/scala/collection/concurrent/CNodeBase.java)
- [scala.beans.ScalaBeanInfo](//github.com/jxcoder/scala/blob/2.11.x/src/library/scala/beans/ScalaBeanInfo.scala)

### CODE REVIEW

**Comment by jxcoder** [comment]
Hello!

@phaller, @axel22, what do you think about this?

**Comment by jxcoder** [comment]
Where to see what exactly failed?

**Comment by mpociecha** [comment]
https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/7770/console

Do you run tests locally at all?

**Comment by jxcoder** [comment]
@mpociecha, big thanks!

I don't understand CI comment.

> synthetic method
>     [mima]    scala$beans$ScalaBeanInfo$$pd()Array[java.beans.PropertyDescriptor] in class
>     [mima]    scala.beans.ScalaBeanInfo does not have a correspondent in new version

I don't change the specified method.

What's wrong?

File: src/library/scala/beans/ScalaBeanInfo.scala [hunk 1]
```diff
@@ -32,15 +31,9 @@ abstract class ScalaBeanInfo(clazz: java.lang.Class[_],
   override def getPropertyDescriptors() = pd
   override def getMethodDescriptors() = md
 
-  // override def getAdditionalBeanInfo() = Array(Introspector getBeanInfo clazz.getSuperclass)
-
   private def init() {
-    var i = 0
```
Comment (axel22) [inline src/library/scala/beans/ScalaBeanInfo.scala:38]: My guess is that this was originally done for performance reasons, but I think it's fine to replace it with a foreach loop over a range.

File: src/library/scala/collection/concurrent/CNodeBase.java [hunk 2]
```diff
@@ -19,15 +19,15 @@
     public volatile int csize = -1;
```
Comment (axel22) [inline src/library/scala/collection/concurrent/CNodeBase.java:19]: Since this changes a file in a completely different package, you should put this, and other `TrieMap`-related changes into a separate pull request.

File: src/library/scala/collection/concurrent/CNodeBase.java [hunk 3]
```diff
@@ -19,15 +19,15 @@
...
 
     public boolean CAS_SIZE(int oldval, int nval) {
-	return updater.compareAndSet(this, oldval, nval);
+        return updater.compareAndSet(this, oldval, nval);
```
Comment (axel22) [inline src/library/scala/collection/concurrent/CNodeBase.java:22]: Was this a TAB replaced with SPACES, or the other way around?

(we should have 4 SPACES per indentation level in Java files)

File: src/library/scala/collection/convert/Wrappers.scala [hunk 4]
```diff
@@ -411,24 +411,21 @@ private[collection] trait Wrappers {
 
     override def size = underlying.size
```
Comment (axel22) [inline src/library/scala/collection/convert/Wrappers.scala:413]: Since this change is logically separate, it should be in a separate pull request.

File: src/library/scala/collection/convert/Wrappers.scala [hunk 5]
```diff
@@ -411,24 +411,21 @@ private[collection] trait Wrappers {
...
 
-    def get(k: String) = {
-      val v = underlying get k
-      if (v != null) Some(v.asInstanceOf[String]) else None
+    def get(k: String): Option[String] = {
+      Option((underlying get k).asInstanceOf[String])
```
Comment (Ichoran) [inline src/library/scala/collection/convert/Wrappers.scala:415]: Are you sure that the `Option(x)` pattern is as efficient as the `if (x != null) Some(x) else None` pattern?  On something as basic as `get`, we _do_ care.

Comment (jxcoder) [inline src/library/scala/collection/convert/Wrappers.scala:415]: @Ichoran,

In [Option.scala](https://github.com/scala/scala/blob/2.11.x/src/library/scala/Option.scala#L25) the same code:

``` scala
if (x == null) None else Some(x)
```

I don't see the reason why it should be slower.

Of course, there are function calling... Is that so expensive?

File: src/library/scala/beans/ScalaBeanInfo.scala (hunk 1 above)
Comment (Ichoran) [inline src/library/scala/beans/ScalaBeanInfo.scala:38]: I think it is probably _not_ fine because we would hope init is lightweight and this could make it quite a bit heavier (it adds three extra object allocations at least nominally: two to create the range and one for the closure).  Perhaps nobody uses beans any more so it doesn't matter?  In any case, it shouldn't be changed without more justification that it doesn't matter performancewise (i.e. that this will not become a bottleneck in any reasonable scenario).

File: src/library/scala/collection/convert/Wrappers.scala (hunk 5 above)
Comment (Ichoran) [inline src/library/scala/collection/convert/Wrappers.scala:415]: Function calling would, if not optimized out by the JVM, probably add about 5% to the execution time of the code, and the JVM is pretty good at optimizing out such things.  But I don't see much reason to change clear, working code into different clear, working code that might be slower without a demonstration that it really isn't.  Does the change let you sneak under the magic JVM inlining budget, for example?

File: src/library/scala/beans/ScalaBeanInfo.scala (hunk 1 above)
Comment (axel22) [inline src/library/scala/beans/ScalaBeanInfo.scala:38]: It's true that three objects additional are created, but in relative terms compared to the `withFilter` object created a couple of lines above, along with the boxed `getMethods` array, the builder object, the resulting `md` collection with the size equal to the number of methods, and the array for `pd`... this may be insignificant.

Also - ranges should be optimized by the compiler anyway, and the whole `foreach` on the range should really be inlined and transformed into a `while` loop with the optimized build (not 100% sure, but I think this used to be the case). <- this was my main point for saying it is probably fine, but I'm ok either way

Comment (Ichoran) [inline src/library/scala/beans/ScalaBeanInfo.scala:38]: Oh how I _wish_ foreach on ranges were inlined and transformed!  This has been proposed over and over, and shot down by Martin for being insufficiently general; implemented in plugins and macros by Olivier Chafik (ScalaCL, Scalaxy) and Paul Phillips and at least one other; and it's still not in.  A simple while loop skipping by 3, even on optimized, turns from a compact local 31-byte method into a hulking 189-byte monstrosity with over a dozen method calls.  Not having to write while loops would save me _so_ much grief.  Anyway, I don't worry about the general case where there are many objects, but about the case where there are very few properties (zero or one), which I suspect is probably quite common.

Comment (Ichoran) [inline src/library/scala/beans/ScalaBeanInfo.scala:38]: I do agree that the previous code may be more expensive, due to the filter etc. on the methods.  Maybe that should be sped up, if this is at all important.

Comment (axel22) [inline src/library/scala/beans/ScalaBeanInfo.scala:38]: Indeed - having a guarantee that range foreach is always transformed would be beautiful.
I thought Miguel also did some work there a few years back.

**Comment by adriaanm** [comment]
Thank you for your contribution, @jxcoder. Your version is more elegant, but, at this time, I think we better leave this code as-is. 

Please don't be discouraged by this! May I suggest focussing your refactoring efforts on the 2.12.x branch, once it picks up steam (after Scaladays SF)? The cost of changing code that's not broken sometimes outweighs the benefits of cleanliness, especially in a maintenance branch such as 2.11.x.

**Comment by jxcoder** [comment]
@adriaanm

What can I refactor right now in 2.11.x?

**Comment by adriaanm** [comment]
@jxcoder, we don't have a list of refactorings we'd like to do. We try to only refactor when necessary to fix bugs/add features. Over time, we've often found the cost of refactoring to be underestimated.

So, while you're very welcome to refactor code, we have to enforce a pretty high bar -- sorry. The benefits of refactoring must significantly outweigh the costs (see below for some random criteria that come to mind).

A great way to make the benefits more apparent is by fixing a bug/improving documentation, and then doing some refactoring work around the area (keep bug fix and refactoring in separate commits).

Benefits:
- that great feeling of knowing the world is a more organized place :-)
- more maintainable code
- more performant code
- more correct code
- better tested code

Costs:
- merges become harder
- risk of introducing subtle regressions
- original author doesn't recognize/understand their code

**Comment by som-snytt** [comment]
> original author doesn't recognize/understand their code
> In six months, no one recognizes their own code.

For me, that happens in about four days.

I think we (excuse the royal we) underestimate the cost of merging refactors (equivalent to the cost of whitespace changes), and also underestimate the utility of abandoned branches of refactorings.

My New Year's resolution is to create one abandoned branch every weekend of OSS refactorings that benefits only my own poor understanding of the code.

And sometimes a refactor leads to real insight that's worth committing.

The cost of merging includes the cost of the eyeballs required to LGTM.
